sync/src/main/java/tech/pegasys/artemis/sync/FetchRecentBlocksService.java

```diff
@@ -0,0 +1,249 @@
+    return blockSubscribers.subscribe(subscriber);
+  }
+
+  public void unsubscribeBlockFetched(final int subscriberId) {
+    blockSubscribers.unsubscribe(subscriberId);
+  }
+
+  private void setupSubscribers() {
+    this.pendingBlocksPool.subscribeRequiredBlockRoot(this::requestRecentBlock);
+    this.pendingBlocksPool.subscribeRequiredBlockRootDropped(this::cancelRecentBlockRequest);
+  }
+
+  public void requestRecentBlock(final Bytes32 blockRoot) {
+    if (pendingBlocksPool.contains(blockRoot)) {
+      // We've already got this block
+      return;
+    }
+    final FetchBlockTask task = fetchBlockTaskFactory.create(eth2Network, blockRoot);
+    if (allTasks.putIfAbsent(blockRoot, task) != null) {
+      // We're already tracking this task
+      task.cancel();
+      return;
+    }
+    LOG.trace("Queue block to be fetched: {}", blockRoot);
+    queueTask(task);
+  }
+
+  public void cancelRecentBlockRequest(final Bytes32 blockRoot) {
+    final FetchBlockTask task = allTasks.get(blockRoot);
+    if (task != null) {
+      task.cancel();
+    }
+  }
+
+  private synchronized void checkTasks() {
+    // If we have capacity, execute the next task
+    if (activeTasks.size() < maxConcurrentRequests) {
+      final FetchBlockTask nextTask = pendingTasks.poll();
+      if (nextTask == null) {
+        return;
+      }
+      runNextTask(nextTask);
+    }
+  }
+
+  private void runNextTask(final FetchBlockTask task) {
+    registerActiveTask(task);
+    task.run()
+        .thenAccept(res -> processFetchResult(task, res))
+        .exceptionally(
+            (err) -> {
+              LOG.warn("Failed to run " + task.getClass().getSimpleName(), err);
+              return null;
+            })
+        .always(() -> deregisterActiveTask(task));
+  }
+
+  private void processFetchResult(final FetchBlockTask task, final FetchBlockResult result) {
+    switch (result.getStatus()) {
+      case SUCCESSFUL:
+        handleFetchedBlock(task, result.getBlock());
+        break;
+      case NO_AVAILABLE_PEERS:
+        // Wait a bit and then requeue
+        queueTaskWithDelay(task, WAIT_FOR_PEERS_DURATION);
+        break;
+      case FETCH_FAILED:
+        // Push task back onto queue to retry
+        queueTaskWithRetryDelay(task);
+        break;
+      case CANCELLED:
+        LOG.trace("Request for block cancelled: {}.", task.getBlockRoot());
+        removeTask(task);
+        break;
+    }
+  }
+
+  private void registerActiveTask(FetchBlockTask task) {
+    LOG.trace("Fetch block {}", task.getBlockRoot());
+    activeTasks.add(task);
+  }
+
+  private void deregisterActiveTask(FetchBlockTask task) {
+    activeTasks.remove(task);
+    checkTasks();
+  }
+
+  private void removeTask(FetchBlockTask task) {
+    // Stop tracking task
+    task.cancel();
+    allTasks.compute(
+        task.getBlockRoot(),
+        (root, existingTask) -> {
+          if (Objects.equals(task, existingTask)) {
```

Comment:
In what scenario here can the if statement return false?

Reply:
And if there can be multiple tasks for each block root, why not have a set instead of a 1 to 1 mapping

Reply:
I don't think this should ever actually return false - its just a safety thing to make sure that if somehow the task had been overwritten with a separate task, we won't interfere with the newer task.  There's a simpler way to do this though:  `allTasks.remove(task.getBlockRoot(), task)`.  Will update.